shared/hashutil/hash.go

```diff
@@ -10,7 +10,15 @@ import (
 func Hash(data []byte) [32]byte {
 	var hash [32]byte
 
-	h := sha3.Sum256(data)
-	copy(hash[:], h[:32])
+	h := sha3.NewLegacyKeccak256()
+
+	// The hash interface never returns an error, for that reason
+	// we are not handling the error below. For reference, it is
+	// stated here https://golang.org/pkg/hash/#Hash
+
+	// #nosec G104
+	h.Write(data)
+	h.Sum(hash[:0])
+
 	return hash
```

Comment:
Can't you just return `sha3.Sum256(data)`? what's the difference here

Reply:
Its actually not the same, there is the legacy keccak 256 which was created before the standards were finalised and there is the currently finalised standard. They give different results:
https://github.com/golang/crypto/blob/master/sha3/hashes.go

Reply:
Hm.. I wonder how we know which one to use

Reply:
This one : `NewLegacyKeccak256()` , the padding in it is the same as the crypto library in go-ethereum. Basically the struct here is : `&state{rate: 136, outputLen: 32, dsbyte: 0x01} ` whereas for the regular sha3 it is `&state{rate: 136, outputLen: 32, dsbyte: 0x06} ` . This is confirmed when you look at the hashing function in geth:
https://github.com/ethereum/go-ethereum/blob/master/crypto/crypto.go#L56
https://github.com/ethereum/go-ethereum/blob/master/crypto/sha3/hashes.go#L16

Reply:
Should we be using the go-ethereum library instead of the golang one? Or does it not matter based on the above?

Reply:
We discussed offline that the go-Ethereum library is expensive to build due to libsepc256k1 so we should avoid it unless absolutely necessary.